rewatch/src/project_context.rs

```diff
@@ -0,0 +1,288 @@
+
+    for dep in dependencies_from_config {
+        // Monorepo packages are expected to be symlinked in node_modules.
+        if let Ok(dep_path) = folder_path
+            .join("node_modules")
+            .join(dep)
+            .canonicalize()
+            .map(helpers::StrippedVerbatimPath::to_stripped_verbatim_path)
+        {
+            debug!(
+                "Checking if dependency \"{}\" is a local package at \"{}\"",
+                dep,
+                dep_path.display()
+            );
+            if helpers::is_local_package(folder_path, &dep_path) {
+                local_dependencies.insert(dep.to_string());
+            }
+        }
+    }
+
+    Ok(local_dependencies)
+}
+
+fn monorepo_or_single_project(
+    path: &Path,
+    current_rescript_json: RescriptJsonPath,
+    current_config: Config,
+) -> Result<ProjectContext> {
+    let local_dependencies = match &current_config.dependencies {
+        None => AHashSet::<String>::new(),
+        Some(deps) => read_local_packages(path, deps)?,
+    };
+    let local_dev_dependencies = match &current_config.dev_dependencies {
+        None => AHashSet::<String>::new(),
+        Some(deps) => read_local_packages(path, deps)?,
+    };
+    if local_dependencies.is_empty() && local_dev_dependencies.is_empty() {
+        Ok(ProjectContext::SingleProject {
+            config: current_config,
+            path: current_rescript_json.clone(),
+        })
+    } else {
+        Ok(ProjectContext::MonorepoRoot {
+            config: current_config,
+            path: current_rescript_json.clone(),
+            local_dependencies,
+            local_dev_dependencies,
+        })
+    }
+}
+
+impl ProjectContext {
+    pub fn new(path: &Path) -> Result<ProjectContext> {
+        let path = helpers::get_abs_path(path);
+        let (current_config, current_rescript_json) = packages::read_config(&path)
+            .map_err(|_| anyhow!("Could not read rescript.json at {}", path.to_string_lossy()))?;
+        let nearest_parent_config_path = match path.parent() {
+            None => Err(anyhow!(
+                "The current path \"{}\" does not have a parent folder",
+                path.to_string_lossy()
+            )),
+            Some(parent) => Ok(helpers::get_nearest_config(parent)),
+        }?;
+        let context = match nearest_parent_config_path {
+            None => monorepo_or_single_project(&path, current_rescript_json, current_config),
+            Some(parent_config_path) => {
+                match packages::read_config(parent_config_path.as_path()) {
+                    Err(e) => Err(anyhow!(
+                        "Could not read the parent config at {}: {}",
+                        parent_config_path.to_string_lossy(),
+                        e
+                    )),
+                    Ok((workspace_config, workspace_rescript_json)) => {
+                        let is_current_config_listed_in_workspace = workspace_config
```

Comment:
I would try to maybe remove the nesting ever so slightly here - nesting multiple matches. Perhaps break out the config reading to another function as well?